cmd/lotus-miner/is_stop_safe.go

```diff
@@ -0,0 +1,318 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	datatransfer "github.com/filecoin-project/go-data-transfer"
+	"github.com/filecoin-project/go-fil-markets/retrievalmarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/dline"
+	"github.com/filecoin-project/lotus/api"
+	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+)
+
+type IsStopSafe struct {
+	MinerAddress        address.Address     `json:"miner-address"`
+	DeadlineFaultCutoff DeadlineFaultCutoff `json:"deadline-fault-cutoff"`
+	Safe                bool                `json:"safe"`
+	Errors              []string            `json:"errors"`
+}
+
+type DeadlineFaultCutoff struct {
+	CurrentEpoch         abi.ChainEpoch `json:"current-epoch"`
+	CutoffEpoch          abi.ChainEpoch `json:"cutoff-epoch"`
+	EstimatedSecondsLeft float32        `json:"estimated-seconds-left"`
+}
+
+var isStopSafeCmd = &cli.Command{
+	Name:  "is-stop-safe",
```

Comment:
I'd make this be 2 sub-commands of the `stop` command:
* `lotus-miner stop check` which just tells you if it's ok to stop
* `lotus-miner stop safe` which stops when it's safe

Reply:
I didn't like the command placement either. Thanks for the direction!

Should `lotus-miner stop safe` subscribe to/poll state changes until it's safe and then stop? If so, then maybe `lotus-miner stop check --wait` should also exist? Or did you have in mind that it would fail immediately if it's not safe.

Either way, it could be argued that `lotus-miner stop` should always perform checks by default and maybe `lotus-miner stop --force` should be the current behavior. This is, however, a breaking UI change.

Reply:
> subscribe to/poll state changes until it's safe and then stop?

Yes, it can just poll changes

> maybe lotus-miner stop check --wait should also exist?

Good idea, and should be easy to add when we have `stop safe` (also, `safe` can just be a `--safe` flag if we go with `--force` bolew)

> it could be argued that lotus-miner stop should always perform checks by default and maybe lotus-miner stop --force should be the current behavior. This is, however, a breaking UI change.

I thought about that, but would prefer to avoid the breaking change for now. Maybe we could:
* Add a warning when normal `stop` is called saying something like 'performing unsafe node shutdown. (NOTE: in near future safe shutdown will become the default. Specify --force to keep current behavior)'
* Not print the warning if `--force` is specified
* At some point in te future (in one or two versions) change the default behavior to safe

Reply:
@magik6k Thinking about this more:

I'm not sure a `wait` feature is actually useful. If you're stopping a miner because you want to take it down permanently then you likely don't care how safe it is. If you're stopping a miner because you plan to upgrade/fix/move it, then you aren't going to want it to stop at an unknown point in the future. You'd want to have planned down time where you are available to do the transition at the most opportune time (hence the check command gives you the time to the next deadline fault cutoff).

What's more, if all we are doing is polling, then it's trivial to re-implement that manually:
```bash
while ! lotus-miner stop --check; do sleep 1; done; lotus-miner stop
```

As such I'd propose we simply move this check into `lotus-miner stop --check` and add a `lotus-miner stop --force` command that does the current behavior. Then `lotus-miner stop` would also do the current behavior with a deprecation warning. In the future `lotus-miner stop` will simply fail if the checks fail. It won't wait.

Thoughts?